File: 049-把字符串转换成整数/049-StrToInt.cpp
class Solution {
public:
    int StrToInt(string str) {
        string::iterator pstr = str.begin( );
        while (*pstr == ' ')
        {
            pstr++;
        }
        bool minus = false;
        if (*pstr == '+')
        {
            pstr++;
        }
        else if (*pstr == '-')
        {
            pstr++;
            minus = true;
        }

        long long int value = 0;
        for (; pstr != str.end( ); pstr++)
        {
            if ('0' <= *pstr && *pstr <= '9')
            {
                value *= 10;
                value += *pstr - '0';
            }
            else
            {
                break;
            }
            if((minus == true  && value > (unsigned long)(INT_MAX) + 1)
            || (minus == false && value > INT_MAX))
            {
                break;
            }

        }
        if(pstr != str.end( ))
        {
            return 0;
        }
        else
        {

            if (minus == true)
            {
                value = -value;
            }

            if (value >= INT_MAX)
            {
                value = INT_MAX;
            }
            else if (value <= INT_MIN)
            {
                value = INT_MIN;
            }

            return (int)value;

        }
    }
};

File: 049-把字符串转换成整数/049-StrToInt.py
# -*- coding:utf-8 -*-
class Solution:
    def StrToInt(self, s):
        # write code here
        if s=='':
            return 0
        s=s.strip()
        valid={'0':0,'1':1,'2':2,'3':3,'4':4,'5':5,'6':6,'7':7,'8':8,'9':9}
        minus=False
        if s[0]=='+':
            s=s[1:]
        elif s[0]=='-':
            minus=True
            s=s[1:]
        elif s[0] not in valid:
            return 0
        ans=0
        for char in s:
            if char in valid:
                ans*=10
                ans+=valid[char]
            else:
                return 0
        if minus:
            ans= -ans
        if -2147483648<=ans <=2147483647:
            return ans
        else:
            return 0
